Add stock location through ps_stock_available table

matks opened the pull request:
⚠️ PR mainly approved, only last commit https://github.com/PrestaShop/PrestaShop/pull/10422/commits/1abf49c9400dff20dac1a76d8ea410084f98f27b needs review

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Restore the use of the 'stock location' product field
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #10113
| How to test?  |The expected behavior is described in the issue

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/10422)
<!-- Reviewable:end -->

matks committed:
Enable field stock location for product and combinations

matks committed:
Display stock location in BO Order page

matks committed:
Handle validation errors for Admin ProductController ajax usecases and refacto sending error JSON responses

mickaelandrieu commented on classes/order/Order.php:
```diff
@@ -699,7 +699,7 @@ protected function setProductCustomizedDatas(&$product, $customized_datas)
      *
      * @param array &$product
      */
-    protected function setProductCurrentStock(&$product)
```
this will break the compatibility: you can create a new protected method and deprecate this one if you think the naming of the former one is bad.

matks replied:
First I considered adding a new protected method and keeping this one but the behavior change is too important, I think it is better to embark it inside the existing function so that developers using it do not have "lose" the feature :) 

Incoming in next commit

mickaelandrieu commented on classes/stock/StockAvailable.php:
```diff
@@ -54,6 +54,9 @@ class StockAvailableCore extends ObjectModel
     public $out_of_stock = false;
 
+    /** @var string the location of the stock for this product / combination */
+    public $location = '';
```
:+1:

mickaelandrieu commented on classes/stock/StockAvailable.php:
```diff
@@ -353,6 +357,44 @@ public static function setProductOutOfStock($id_product, $out_of_stock = false,
+     */
+    public static function setLocation($id_product, $location, $id_shop = null, $id_product_attribute = 0)
+    {
+        if (false === Validate::isUnsignedId($id_product)) {
```
We can do better here:

```php
if (
    false === Validate::isUnsignedId($id_product)
    || false === Validate::isUnsignedId($id_shop)
   || false === Validate::isUnsignedId($id_product_attribute)
   || false === Validate::isString($location)
) {
    throw new SomeSpecializedException('Some meaningful message with serialized data passed');
}
```

And you can remove almost all cast operations you're doing, also this will remove unexpected weird behaviors that can come from wrong data passed to this function.

mickaelandrieu commented on classes/stock/StockAvailable.php:
```diff
@@ -698,6 +740,33 @@ public static function outOfStock($id_product, $id_shop = null)
+    public static function getLocation($id_product, $id_product_attribute = null, $id_shop = null)
+    {
+        if (false === Validate::isUnsignedId($id_product)) {
+            return false;
```
Same here, I think we should throw an exception instead :) it can be a simple InvalidArgumentException but avoid missing multiple return types to manage wrong inputs. Also, you can validate other fields.

mickaelandrieu commented on classes/stock/StockAvailable.php:
```diff
@@ -698,6 +740,33 @@ public static function outOfStock($id_product, $id_shop = null)
+     *
+     * @return bool|string
+     */
+    public static function getLocation($id_product, $id_product_attribute = null, $id_shop = null)
```
$id_shop should be default evaluated to 0, same for $id_product_attribute.

matks replied:
Would it introduce some magic numbers inside this function ? I know in PrestaShop ID 0 is the default one. However adding it as default value would hard-code this rule into the code. What if one day we want to move to uuid SQL IDs ? Even if we extend StockAvailable, the value `0` will be hardcoded into the function signature.

So ... what about `null` as default value always :) ?

mickaelandrieu replied:
if we move from ids to uuid we will have bigger problems as we expected integer or null here and we will change the functions to use some Uuid objects, isn't it?

It's only 1 implementation, I don't think we will break it in 1.7.x release.

Anytime, when we can do it I want only 1 possible type of input (if we were able to use PHP 7.1 already, I could be fine with ?int but ... :'( ).

Other option: check if $id_product_attribute is a int if it's not a null value, wdyt?

PierreRambaud replied:
Even if we use only one type, we are able to set a value to `null`. Which is logic to have when there is no id attached to the object, because 0 is an id :)
If we move to uuid or anything else, it's easier to test if value is null instead of checking if it's numeric and not 0 (and it costs more resources).

```php
<?php

class A
{
    public function __construct(int $a)
    {
        var_dump($a);
    }
}

new A();
new A(22);
new A('azerty');
```

Will produce:
```bash
$ php7.2 test.php 
/home/prambaud/test.php:7:
NULL
/home/prambaud/test.php:7:
int(22)
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to A::__construct() must be of the type integer or null, string given, called in /home/prambaud/test.php on line 13 and defined in /home/prambaud/test.php:5
Stack trace:
#0 /home/prambaud/test.php(13): A->__construct('azerty')
#1 {main}
  thrown in /home/prambaud/test.php on line 5
```

mickaelandrieu commented on classes/stock/StockAvailable.php:
```diff
@@ -698,6 +740,33 @@ public static function outOfStock($id_product, $id_shop = null)
+        $query->from('stock_available');
+        $query->where('id_product = ' . (int) $id_product);
+        $query->where('id_product_attribute = ' . (int) $id_product_attribute);
+
```
here, you can do the trick to stick with the old API:

```php
// StockAvailable::addSqlShopRestriction expects Shop id to be null if not valid.
if ($id_shop !== 0) {
    $id_shop = null;
}
```

mickaelandrieu commented on src/Adapter/Product/AdminProductWrapper.php:
```diff
@@ -215,6 +216,15 @@ public function processProductOutOfStock(Product $product, $out_of_stock)
+     */
+    public function processLocation(Product $product, $location)
+    {
+        StockAvailable::setLocation((int) $product->id, $location);
```
you don't need to cast to integer if it comes from Product.

mickaelandrieu commented on src/Adapter/Product/ProductDataProvider.php:
```diff
@@ -111,6 +112,17 @@ public function getQuantity($id_product, $id_product_attribute = null, $cache_is
+     *
+     * @return string
+     */
+    public function getLocation($id_product, $id_product_attribute = null)
```
I think you should use a default int value instead, as we already know 0 is the default value :+1:

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/ProductController.php:
```diff
@@ -1130,4 +1156,19 @@ public function renderFieldAction($productId, $step, $fieldName)
+     *
+     * @return JsonResponse
+     */
+    private function returnErrorJsonResponse(array $errors, $httpStatusCode)
```
how about moving it into the parent class?

matks committed:
Improve error and validation from PR feedback

matks commented:
@mickaelandrieu PR feedbacks have been implemented in https://github.com/PrestaShop/PrestaShop/pull/10422/commits/5b87438a839613641fb63f48f6a17fabd791fffa ;)

eternoendless commented:
Code looks good, but tests are failing

matks committed:
Comply to unit tests requirements

matks commented:
@eternoendless I saw 😢 I fixed it by removing some validations in `StockAvailable`

According to [AdminModelAdapterTest](https://github.com/PrestaShop/PrestaShop/blob/develop/tests/PrestaShopBundle/Model/Product/AdminModelAdapterTest.php#L136) we must accept `string` values for `$id_product_attribute` and `$id_product` in [getLocation()](https://github.com/PrestaShop/PrestaShop/pull/10422/files#diff-6a810cff1b3b082a1b2a3f1dbca30eeeR764)

Fixed by https://github.com/PrestaShop/PrestaShop/pull/10422/commits/9eae1b1267fb1093778bdbb02e2c9a3ba67c930d, should fix Travis build

mickaelandrieu commented on classes/stock/StockAvailable.php:
```diff
@@ -353,6 +357,58 @@ public static function setProductOutOfStock($id_product, $out_of_stock = false,
+                'id_product_attribute' => $id_product_attribute,
+                'location' => $location,
+            ];
+            throw new \InvalidArgumentException(sprintf(
```
isn't weird to throw an exception if the data is actually valid?

mbadrani commented:
when I fill stock location like this

![image](https://user-images.githubusercontent.com/29543279/45768593-985c1a80-bc3d-11e8-9034-7bcffcbcf575.png)

... And I access to order page, it display the number of quantity instead of stock location value

![image](https://user-images.githubusercontent.com/29543279/45768546-7cf10f80-bc3d-11e8-99ef-0c1bdc8ccde6.png)

matks commented:
@mbadrani Can you clear your cache on this very page (in chrome, open "developer tools", hold click on the "refresh page button" to make a menu appear and then select "clear cache and perform a forced actualisation of the page" ?

On your screenshot I see the row elements are dashed so I think is a an outdated CSS issue

mbadrani commented:
After force refreshing and clear the cache I can reproduce this weired behaviour

matks committed:
Handle customized products usecase for stock location display in order

matks commented:
@mbadrani I found an issue with customized data indeed 😉 nice catch !

⚠️ Last commit https://github.com/PrestaShop/PrestaShop/pull/10422/commits/1abf49c9400dff20dac1a76d8ea410084f98f27b should fix it, waiting for review now

jolelievre reviewed (dismissed):
good job!

mbadrani commented:
@matks After this fix I have another bug below
https://drive.google.com/file/d/1eB-OvunAIy-46uiOKDtODnrJg8FBUyZ3/view?usp=sharing

PierreRambaud commented:
I think it's not a bug due to his modification because I sometimes have the same bug on develop.

Quetzacoalt91 commented:
We never know, if you look at the symfony debug bar on the video, it seems an error 500 occurred (the icon becomes red).

@mbadrani can you please hover this icon and click on the sha1 related to the call in error? You will get the details.

matks committed:
Handle all exceptions in ProductController for ajax requests failures

matks committed:
Handled bad form submission usecase for empty stock location field

matks commented:
Again nice catch @mbadrani : the bug happens if you empty the "stock location" field. This was considered a not valid dataset by the code. I fixed:
- the usecase
- now even if the dataset is not valid, the BO does not freeze the way it does in your video

**Technical insights:**
Because of https://github.com/PrestaShop/PrestaShop/pull/10422/commits/5b87438a839613641fb63f48f6a17fabd791fffa#diff-6a810cff1b3b082a1b2a3f1dbca30eeeR384 `null` data was not accepted anymore as a valid input and was consequently rejected using a `\InvalidArgumentException`. However since that was an ajax call, JSON was expected although it was the HTML symfony error page that was returned, causing JS to crash.

mbadrani commented:
@PierreRambaud it's not a random bug it has been easily reproduced by @matks

jolelievre reviewed (approved):
nice fix!

mbadrani commented:
Well done guys!

eternoendless commented:
Woot! 💪 
Let's merge this once travis is green

Quetzacoalt91 commented:
Thank you @matks

123monsite-regis commented:
hi,

about this, can we undestand that "public $location;" inside Combination Object definition is now obsolete and should be migrated when updating shop ?
i talk about : https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Combination.php#L40
`    public $location;`


in fact some paid modules still use this field whith the same label, it's quite confusing.


regards
